### ISSUE TITLE
Calling plot(m, forecast) freezes my computer (duplicate dates in history)

### ISSUE THREAD

**Issue by danmaxis** [facebook/prophet#1476]
Hello there,
I'm having a serious issue using prophet (R version) on Windows 10. When I call the plot function, my computer freezes pretty badly. Didnt have this issue using the library on linux.

My setup:
- Processor: AMD Ryzen 5 1600 3.2Ghz
- RAM: 16GB
- OS: Windows 10 1909 18363.815
- R version: 3.6.3
- R Studio version: 1.2.5042
- Prophet reference version: both and master 
- Rbuildtools: 3.5

I've also set those rstan vars for successful compilation: [](https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started)

Executing the following R code
```
m <- prophet(cymt %>% ungroup())
future <- make_future_dataframe(m, freq="year",periods = 3)
tail(future)
forecast <- predict(m, future)
plot(m, forecast)
```

I've tested with other plot calls unrelated to prophet and they worked normally.
I would be glad to post more information if it is needed.

Some windows log entries before freezing:

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="ESENT" />
  <EventID Qualifiers="0">455</EventID>
  <Level>2</Level>
  <Task>3</Task>
  <Keywords>0x80000000000000</Keywords>
  <TimeCreated SystemTime="2020-05-04T17:40:56.612338400Z" />
  <EventRecordID>62711</EventRecordID>
  <Channel>Application</Channel>
  <Computer>Computer</Computer>
  <Security />
  </System>
- <EventData>
  <Data>svchost</Data>
  <Data>6892,R,98</Data>
  <Data>TILEREPOSITORYS-1-5-18:</Data>
  <Data>C:\WINDOWS\system32\config\systemprofile\AppData\Local\TileDataLayer\Database\EDB.log</Data>
  <Data>-1023 (0xfffffc01)</Data>
  </EventData>
  </Event>    
```

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="EventLog" />
  <EventID Qualifiers="32768">6008</EventID>
  <Level>2</Level>
  <Task>0</Task>
  <Keywords>0x80000000000000</Keywords>
  <TimeCreated SystemTime="2020-05-04T17:33:38.362506100Z" />
  <EventRecordID>38427</EventRecordID>
  <Channel>System</Channel>
  <Computer>Computer</Computer>
  <Security />
  </System>
- <EventData>
  <Data>14:02:23</Data>
  <Data>‎04/‎05/‎2020</Data>
  <Data />
  <Data />
  <Data>4809</Data>
  <Data />
  <Data />
  <Binary>E4070500010004000E0002001700EB01E407050001000400110002001700EB01600900003C000000010000006009000001000000B00400000000000000000000</Binary>
  </EventData>
  </Event>
```

```
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="Microsoft-Windows-DistributedCOM" Guid="{1B562E86-B7AA-4131-BADC-B6F3A001407E}" EventSourceName="DCOM" />
  <EventID Qualifiers="0">10016</EventID>
  <Version>0</Version>
  <Level>3</Level>
  <Task>0</Task>
  <Opcode>0</Opcode>
  <Keywords>0x8080000000000000</Keywords>
  <TimeCreated SystemTime="2020-05-04T17:37:09.236457900Z" />
  <EventRecordID>38486</EventRecordID>
  <Correlation ActivityID="{2c23045f-a77c-42a8-ac6c-cfd6d6be9c46}" />
  <Execution ProcessID="1052" ThreadID="12324" />
  <Channel>System</Channel>
  <Computer> Computer </Computer>
  <Security UserID="S-1-5-21-546434611-2804603848-1334593372-1001" />
  </System>
- <EventData>
  <Data Name="param1">app specific</Data>
  <Data Name="param2">Place</Data>
  <Data Name="param3">Activation</Data>
  <Data Name="param4">{2593F8B9-4EAF-457C-B68A-50F6B8EA6B54}</Data>
  <Data Name="param5">{15C20B67-12E7-4BB6-92BB-7AFF07997402}</Data>
  <Data Name="param6"> Computer </Data>
  <Data Name="param7">user</Data>
  <Data Name="param8">S-1-5-21-546434611-2804603848-1334593372-1001</Data>
  <Data Name="param9">LocalHost (Using LRPC)</Data>
  <Data Name="param10">Not Available</Data>
  <Data Name="param11"> Not Available </Data>
  </EventData>
  </Event>
```

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="Microsoft-Windows-DistributedCOM" Guid="{1B562E86-B7AA-4131-BADC-B6F3A001407E}" EventSourceName="DCOM" />
  <EventID Qualifiers="0">10010</EventID>
  <Version>0</Version>
  <Level>2</Level>
  <Task>0</Task>
  <Opcode>0</Opcode>
  <Keywords>0x8080000000000000</Keywords>
  <TimeCreated SystemTime="2020-05-04T17:30:48.367276100Z" />
  <EventRecordID>38417</EventRecordID>
  <Correlation ActivityID="{5e31cd65-2ee0-4854-81b6-46553f774327}" />
  <Execution ProcessID="1068" ThreadID="14560" />
  <Channel>System</Channel>
  <Computer>Computer</Computer>
  <Security UserID="S-1-5-21-546434611-2804603848-1334593372-1001" />
  </System>
- <EventData>
  <Data Name="param1">ClevLab.FeedLab_3.1.4.0_x64__qdcg6xvbhrn16!App.AppXe61h5bh74tywv2rz6s4pz28fte7vz70y.mca</Data>
  </EventData>
  </Event>

```

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="Microsoft-Windows-DistributedCOM" Guid="{1B562E86-B7AA-4131-BADC-B6F3A001407E}" EventSourceName="DCOM" />
  <EventID Qualifiers="0">10010</EventID>
  <Version>0</Version>
  <Level>2</Level>
  <Task>0</Task>
  <Opcode>0</Opcode>
  <Keywords>0x8080000000000000</Keywords>
  <TimeCreated SystemTime="2020-05-04T17:29:23.302229300Z" />
  <EventRecordID>38416</EventRecordID>
  <Correlation ActivityID="{5b3e0fc6-5701-4a5c-90e8-39468a826225}" />
  <Execution ProcessID="1068" ThreadID="8608" />
  <Channel>System</Channel>
  <Computer>Computer</Computer>
  <Security UserID="S-1-5-21-546434611-2804603848-1334593372-1001" />
  </System>
- <EventData>
  <Data Name="param1">Microsoft.Todos_2.16.31063.0_x64__8wekyb3d8bbwe!App.AppXcakmdstz39gy6mq7cx2par6qckax124x.mca</Data>
  </EventData>
  </Event>
```

**Comment by bletham** [comment]
This is pretty surprising since the plot code is pretty simple and doesn't do anything that I'd expect to be very computationally intensive. Unless maybe you have a very large number of points in the history? How many rows does `cymt %>% ungroup()` have?

Perhaps the best way to debug this would be to go through the plot call line-by-line to see the particular line that produces the issue (it's not much code):
```
df <- prophet:::df_for_plotting(m, forecast)

gg <- ggplot2::ggplot(df, ggplot2::aes(x = ds, y = y)) +
    ggplot2::labs(x = 'ds', y = 'y')

gg <- gg +
      ggplot2::geom_ribbon(
        ggplot2::aes(ymin = yhat_lower, ymax = yhat_upper),
        alpha = 0.2,
        fill = "#0072B2",
        na.rm = TRUE
      )

gg <- gg +
    ggplot2::geom_point(na.rm=TRUE) +
    ggplot2::geom_line(
        ggplot2::aes(y = yhat), color = "#0072B2",
         na.rm = TRUE) +
    ggplot2::theme(aspect.ratio = 3 / 5)
```

And it sounds like you checked this already, but we just pushed a new version to CRAN last week so do make sure you've re-installed since then and are on 0.6.1.

**Comment by danmaxis** [comment]
Hello there,
`cymt%>% ungroup` is like this:
```
tibble [7,455 x 3] (S3: tbl_df/tbl/data.frame)
 $ ds  : Date[1:7455], format: "2000-01-01" "2000-02-01" "2000-02-01" "2000-02-01" ...
 $ type: chr [1:7455] "Agravo de Instrumento" "A\\xc3\\xa7\\xc3\\xa3o Rescis\\xc3\\xb3ria" "Agravo" "Agravo de Instrumento" ...
 $ y   : int [1:7455] 12 1 4 35 2 151 1 25 2 1
```

I'm running this exposed code and check if manage to get any good results.

**Comment by bletham** [comment]
OK, the issue is definitely the 274,226 rows returned by `df_for_plotting`, ggplot is going to struggle to render that many points on the plot. Given only 7,455 rows in the input dataframe and only 3 future forecast points, I'm not sure what's happening to produce 274,226 rows. `df_for_plotting` does an outer join of the history (which should have 7,455 rows) and the forecast (which should have 7,458 rows):
https://github.com/facebook/prophet/blob/9168dcf11dee8171fb0fbf6a5cf6332e90b97bba/R/R/plot.R#L13-L20

So I'd expect this to have 7,458 rows. Could you verify those counts? What is
```
nrow(m$history)
nrow(future)
nrow(forecast)
```

**Comment by danmaxis** [comment]
Sorry, for taking so long to follow up. You are correct about these expected values.

These are the results you asked:
```
> nrow(m$history)
[1] 7455
> nrow(future)
[1] 7458
> nrow(forecast)
[1] 7458
```

**Comment by bletham** [comment]
hm this code is really producing a dataframe with 274k rows?
```
   df <- m$history %>% 
     dplyr::select(ds, y) %>% 
     dplyr::full_join(forecast, by = "ds") %>% 
     dplyr::arrange(ds) 
```
what are the columns of forecast?

**Comment by danmaxis** [comment]
> hm this code is really producing a dataframe with 274k rows?

Yes, this is the structure:
```
> str(df)
tibble [274,226 x 17] (S3: tbl_df/tbl/data.frame)
 $ ds                        : POSIXct[1:274226], format: "2000-01-01" "2000-02-01" ...
 $ y                         : num [1:274226] 12 1 1 1 1 1 1 1 1 1 ...
 $ trend                     : num [1:274226] 97.4 105.2 105.2 105.2 105.2 ...
 $ additive_terms            : num [1:274226] -395 -133 -133 -133 -133 ...
 $ additive_terms_lower      : num [1:274226] -395 -133 -133 -133 -133 ...
 $ additive_terms_upper      : num [1:274226] -395 -133 -133 -133 -133 ...
 $ yearly                    : num [1:274226] -395 -133 -133 -133 -133 ...
 $ yearly_lower              : num [1:274226] -395 -133 -133 -133 -133 ...
 $ yearly_upper              : num [1:274226] -395 -133 -133 -133 -133 ...
 $ multiplicative_terms      : num [1:274226] 0 0 0 0 0 0 0 0 0 0 ...
 $ multiplicative_terms_lower: num [1:274226] 0 0 0 0 0 0 0 0 0 0 ...
 $ multiplicative_terms_upper: num [1:274226] 0 0 0 0 0 0 0 0 0 0 ...
 $ yhat_lower                : num [1:274226] -2894 -2641 -2525 -2456 -2593 ...
 $ yhat_upper                : num [1:274226] 2267 2364 2551 2573 2540 ...
 $ trend_lower               : num [1:274226] 97.4 105.2 105.2 105.2 105.2 ...
 $ trend_upper               : num [1:274226] 97.4 105.2 105.2 105.2 105.2 ...
 $ yhat                      : num [1:274226] -297.3 -28.3 -28.3 -28.3 -28.3 ... 
```

> what are the columns of forecast?
```
> str(forecast)
'data.frame':	7458 obs. of  16 variables:
 $ ds                        : POSIXct, format: "2000-01-01" "2000-02-01" ...
 $ trend                     : num  97.4 105.2 105.2 105.2 105.2 ...
 $ additive_terms            : num  -395 -133 -133 -133 -133 ...
 $ additive_terms_lower      : num  -395 -133 -133 -133 -133 ...
 $ additive_terms_upper      : num  -395 -133 -133 -133 -133 ...
 $ yearly                    : num  -395 -133 -133 -133 -133 ...
 $ yearly_lower              : num  -395 -133 -133 -133 -133 ...
 $ yearly_upper              : num  -395 -133 -133 -133 -133 ...
 $ multiplicative_terms      : num  0 0 0 0 0 0 0 0 0 0 ...
 $ multiplicative_terms_lower: num  0 0 0 0 0 0 0 0 0 0 ...
 $ multiplicative_terms_upper: num  0 0 0 0 0 0 0 0 0 0 ...
 $ yhat_lower                : num  -2894 -2641 -2525 -2456 -2593 ...
 $ yhat_upper                : num  2267 2364 2551 2573 2540 ...
 $ trend_lower               : num  97.4 105.2 105.2 105.2 105.2 ...
 $ trend_upper               : num  97.4 105.2 105.2 105.2 105.2 ...
 $ yhat                      : num  -297.3 -28.3 -28.3 -28.3 -28.3 ...
```

**Comment by bletham** [comment]
Is there any chance you could share CSVs of m$history and forecast? something weird is happening in the join but I don't have any good guesses for what it might be.

**Comment by danmaxis** [comment]
Sure, this data is from my teacher's open source project [filipezabala/jurimetrics](https://github.com/filipezabala/jurimetrics) .


[m_history.csv](https://github.com/facebook/prophet/files/4632474/m_history.csv.txt)
[forecast.csv](https://github.com/facebook/prophet/files/4632475/forecast.csv.txt)

**Comment by bletham** [comment]
Thanks, I see what's happening here now. The problem is that for some dates (like "2000-02-01") there are multiple rows in the history with that date, but with different values of y. This is because of the different types.
```
> head(m$history)
          ds                                       type   y floor           t
1 2000-01-01                      Agravo de Instrumento  12     0 0.000000000
2 2000-02-01 A\\xc3\\xa7\\xc3\\xa3o Rescis\\xc3\\xb3ria   1     0 0.004737164
3 2000-02-01                                     Agravo   4     0 0.004737164
4 2000-02-01                      Agravo de Instrumento  35     0 0.004737164
5 2000-02-01                          Agravo Regimental   2     0 0.004737164
6 2000-02-01  Apela\\xc3\\xa7\\xc3\\xa3o C\\xc3\\xadvel 151     0 0.004737164
      y_scaled
1 9.035464e-04
2 7.529553e-05
3 3.011821e-04
4 2.635344e-03
5 1.505911e-04
6 1.136963e-02
```
Now this isn't an issue for the model in principle - it can handle having repeated observations at a point. However, using `make_future_dataframe` will by default include all of the history dates, and so has these duplicates also. In the `forecast` dataframe we can see that the model is being evaluated on these repeated points:
```
> head(forecast) %>% select(ds, yhat)
          ds       yhat
1 2000-01-01 -297.29830
2 2000-02-01  -28.25923
3 2000-02-01  -28.25923
4 2000-02-01  -28.25923
5 2000-02-01  -28.25923
6 2000-02-01  -28.25923
```
and so then when we do an outer join between these data frames on ds, it explodes out all of the combinations of these repeated entries and that causes the issue.

This is a bug; things should work with repeated data. So thanks for pointing this out!

That said, in this particular instance - I do question if it makes sense to be combining these observations that seem to be of a different type. What the model will forecast is something like the average over the types. In that case you might as well just average them before passing them into the history (which would allow you to correctly handle situations like the first date where not all types are represented).

**Comment by bletham** [comment]
For the bug...

**Reproduce:**
Here's code to reproduce the issue on the quickstart example:
```
library(prophet)
df <- read.csv('../examples/example_wp_log_peyton_manning.csv')
m <- prophet(rbind(df, df))
print(nrow(df))  # 2905
print(nrow(m$history))  # 5810

future <- make_future_dataframe(m, periods = 3)
forecast <- predict(m, future)
print(nrow(forecast))  # 5813

df <- prophet:::df_for_plotting(m, forecast)
print(nrow(df))  # 11623 ; should be 5813
```

**Workaround:**
Removing the duplicate dates from the `future` dataframe will resolve the issue.
```
future <- make_future_dataframe(m, periods = 3)
future <- data.frame(ds=unique(future$ds))
forecast <- predict(m, future)
print(nrow(forecast))  # 2908

df <- prophet:::df_for_plotting(m, forecast)
print(nrow(df))  # 5813; correct.
```

**The fix:**
`make_future_dataframe` should de-duplicate the dates.
I expect this issue exists in both R and Python.

**Comment by bletham** [comment]
This is fixed in fb65521 .

**Comment by bletham** [comment]
Fix has been pushed to pypi and CRAN in the v1.0 release
